### ISSUE TITLE
Don't warn FS3559 when obj is inferred if the obj is expected there

### ISSUE THREAD

**Issue by cmeeren** [dotnet/fsharp#16454]
The new FS3559 triggers where it should not. Consider this:

```f#
let foo (x: obj) = ()
foo null
```

It triggers on `null`. But the type of the argument is explicitly `obj`, so it doesn't make sense that it should trigger here.

Another slightly more realistic example:

```f#
let invokeStatic (mi: System.Reflection.MethodInfo) =
    mi.Invoke(null, [||])
```

`MethodInfo.Invoke` expects `obj` as its first argument, so F# inferring `obj` here should not cause a warning.

**Comment by Smaug123** [comment]
Yeah, a verbatim `null` is one of the most annoying cases. I tried to fix this and could not :( which is not to say it's impossible, but the required information did not appear to be present locally at the point where the warning is generated.

(The second most annoying case is `typeof<Foo<_>>`, where the `_` is inferred as `obj`.)
